### ISSUE TITLE
Support minimum spike size (s_min) for p=2

### ISSUE THREAD

**Issue by GarrettBlair** [flatironinstitute/CaImAn#747]
**1. Operating system (Linux/macOS/Windows):**  Windows 10 64 bit, 128GB RAM, no GPU, running in parallel on AMD Ryzen 1950x

 **2. Python version (3.x):**  Python ver 3.7.6

 **3. Working environment (Python IDE/Jupyter Notebook/other):** Jupyter Notebook

 **4. Which of the demo scripts you're using for your analysis (if applicable):** demo_pipeline_cnmfe

 **5. CaImAn version:**  1.8.1

 **6. CaImAn installation process (`pip install .`/`pip install -e .`/conda):** pip install

- **Describe the issue that you are experiencing**

Caiman runs very well, no issues with spatial or temporal demixing. But the temporal deconvolution does not seem to be working well. I’ve been using oasis, and varying the temporal parameters don’t seem to affect the outcome (listed below), though I can see the change logged properly in the log file after altering the opts.change_params dict. Varying s_min all had the same output within S after completely re-running cnmf, even at silly thresholds, which is was very perplexing. 

I’m using gcamp7s, so the transients are fairly long compared to the 6f I’ve used before, but this is my first time using caiman (previously used cnmfe in matlab). Below is an image of a typical example, where red dots are the time points of S>0, plotted on top of [C+residuals]. X axis is frames, 10fps

The typical parameter set used, mostly the defaults:  
![typical parameters2](https://user-images.githubusercontent.com/40305871/81136951-d6d94280-8f11-11ea-9dfb-40297ec8d286.png)

So far I have tried the following (in different iterations):
-	Varying s_min: [None, -5, -2, 2, 5, 50] 
-	p = 2
-	Setting optimize_g =True
-	method_deconvolution = ‘oasis’ # I tried ‘cvx’ after installing the two dependencies, but had another issue so I dropped it. ‘cvxpy’ didn’t work either
-	decay_time: [.9, 1, 2, 100]
-	simultaneously = True
-	low_rank_backgroud = True
-	gnb = [0, -1]

Any ideas or recommendation on what I should try are very welcome, I'm out of ideas for now. I'm still learning python honestly, but have a lot of experience in matlab. I ended up writing a silly peak finding method in matlab to work with for now, but using deconvolution just seems more appropriate. Thanks for any help!

-Garret J. Blair

**Comment by j-friedrich** [comment]
Hi Garret,
you didn't attach the image of a typical example, thus I can't tell what's wrong or odd.
For an explanation of the various options oasis has see the [demo](https://github.com/j-friedrich/OASIS/blob/master/examples/Demo.ipynb)
Although it can slightly help to deconvolve/denoise simultaneously, you don't need to rerun the entire pipeline to figure out good parameters for the deconvolution step. Just apply the deconvolve function at the end to the obtained `C+R`

**Comment by GarrettBlair** [comment]
Thank for your response Johannes. Sorry about the omission, here's the example. I'll look into that demo, that's really helpful!
![gjb_deconvolution_alt](https://user-images.githubusercontent.com/40305871/81197342-fc4b6800-8f74-11ea-9fd3-06107bcb87be.png)

**Comment by GarrettBlair** [comment]
I've been testing the oasis package alone, as well as the caiman implementation, and I think I've pinned down why s_min isn't having any effect. 

When using cnmf.fit, while when fitting the AR1 model (p=1), it imports **constrained_oasisAR1** from caiman/source_extraction/cnmf/**oasis.pyx**, which does have the s_min parameter passable.

However, when using the AR2 model (p=2), "constrained_oasisAR2" is called in via the deconvolve function, and s_min is not passed as an argument. Instead, s_min is calculated via the distribution of spike sizes:
- in caiman/source_extraction/cnmf/`deconvolution.py`
- within the definition of `constrained_oasisAR2`
line 901: `spikesizes = np.sort(s[s > 1e-6]);`
line 906: `s_min = spikesizes[i]`
line 907: `tmp = c4smin(y - b, s, s_min)`
Thus why these plots generated via `caiman.source_extraction.cnmf.deconvolution.constrained_foopsi` all look identical, even with very different s_min given (using p=2)
![image](https://user-images.githubusercontent.com/40305871/81348015-259dee00-9072-11ea-9fc0-ac400057084e.png)


I'm guessing this is just a hold over that should be updated and allow s_min to be passed to `deconvolve.constrained.oasisAR2`. My copy of caiman is fairly recent, but let me know if I'm mistaken

**Comment by j-friedrich** [comment]
The implementation of a user provided threshold on the spike size is indeed quite limited for the case p=2 right now in both repos, leaving some work for the future. For the case p=1 it's handeled in caiman's oasis.pyx.

For the moment I have 2 possible suggestions:
Instead of the automatic logarithmic search for `s_min` at the end of `deconvolve.constrained_oasisAR2` you could adapt the code and pass your `s_min` as argument to use that, i.e.  `c = c4smin(y - b, s, s_min)` instead of lines 901-918.

Alternatively, you can use the values `c, s, b, g, lam` returned by `deconvolve.constrained_oasisAR2` and run 
`oasisAR2(y-b, g[0], g[1], s_min=s_min)` from the standalone oasis package, where `y` is your data and `s_min` your threshold. (Setting `jitter=True` might further improve the result slightly.) The latter is a fast but greedy method that is not used in caiman.

**Comment by j-friedrich** [comment]
Solved in the recent 1.8.5 release.
